alf/utils/distributed.py

```diff
@@ -0,0 +1,132 @@
+# Copyright (c) 2021 Horizon Robotics and ALF Contributors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Any, Callable
+import functools
+
+import torch
+from torch.nn.parallel import DistributedDataParallel as DDP
+
+
+class _MethodPerformer(torch.nn.Module):
+    """A nn.Module wrapper whose forward() performs a specified method of
+    the wrapped module.
+
+    The end goal is to make a TARGET METHOD data distributed.
+
+    We need this delegation so that DDP can then wrap over this module. When DDP
+    hijacks the forward() of _MethodPerformer to inject synchronization hooks,
+    it effectively does so for the target method of the wrapped module.
+
+    """
+
+    def __init__(self, module: torch.nn.Module, perform: Callable[..., Any]):
+        """Constructs a _MethodPerformer.
+
+        Args:
+        
+            module: an instance of the module whose method is going to be
+                delegated to. The _MethodPerformer instance needs to access and
+                inherit the parameters from the module, so that DDP knows what
+                parameters to cover.
+
+            perform: the target method of the module. 
+
+        """
+        super().__init__()
+
+        self._wrapped_module = module  # Register and inherit the parameters
+        self._perform = functools.partial(perform, self._wrapped_module)
+
+        # DDP will panic if the wrapped module has member in its state_dict()
+        # that is not a Tensor. Here such state_dict members are picked and
+        # thrown into _ddp_params_and_buffers_to_ignore. By contract this
+        # implicitly instruct DDP wrapper to not include them in its
+        # parameter/buffer synchronization.
+        self._ddp_params_and_buffers_to_ignore = []
+        for name, value in self.state_dict().items():
+            if type(value) is not torch.Tensor:
+                self._ddp_params_and_buffers_to_ignore.append(name)
+
+    def forward(self, *args, **kwargs):
+        return self._perform(*args, **kwargs)
+
+
+def data_distributed(method):
+    """This decorator makes a target method of a module capable of being data
+    distributed via DDP.
+
+    This is to provide a simple and transparent way to enable DDP for specific
+    code logics.
```

Comment:
It will be helpful for the user to use this correctly by noting what is synchronized at when

Reply:
Good point. I will update the documentation.